content/post/20200426-kubernetes-the-universal-control-plane/index.md

```diff
@@ -0,0 +1,113 @@
+
+Kubernetes faces the same challenges as cloud providers do as an orchestrator for containers and infrastructure. It is unsuprising then that Kubernetes also heavily bases its design around declarative APIs as the core user interface.
+
+## First Class Abstractions
+
+Declarative APIs are the defining factor of why cloud providers are attractive on the usability side. Architecting infrastructure this way **reduces friction**: it is simply magnitudes easier for both the provider to build and the user to consume. With the reduced friction, new abstractions in the form of higher level services and primitives like compute schedulers are now commonplace where they used be limited to sophisticated software organizations.
+
+The other side of the coin is that cloud computing offerings, the core primitives we use to build our own services, are locked behind proprietary APIs. Criticizing cloud vendors as proprietary seems antithetical to the concepts described earlier. Even if the APIs to interact with the cloud are proprietary, that certainly does not stop users from creating their own abstractions with cloud compute/storage as the foundation. And one of the benefits of declarative systems is that much of the complexity of infrastructure provisioning is pushed under the surface. The point I am making is more subtle. Although users can build on top of a cloud provider, they can never build abstractions that are **first class** like a cloud provider can.
+
+<!-- Analogy: Building scaffolding; AWS Engineer vs User; Kubernetes Resource Model; Tie in 2 key concepts -->
+To drive the point home, I like to contrast the experience building infrastructure as an AWS engineer vs an engineer that uses AWS. Yes, high level services like Lambda build on top of foundational services like EC2. But there is more to building infrastructure than calling APIs. It comes down to the architecture, the design and practices that AWS built over time to deliver multi-tenant infrastructure at scale. We can see the results in techniques such as [shuffle sharding and static availability](https://aws.amazon.com/builders-library/?cards-body.sort-by=item.additionalFields.customSort&cards-body.sort-order=asc&awsf.filter-content-category=content-category%23architecture) as well as designing services to be managed through control planes. From the outside, we may be able to use the same services but we do not have the same tooling and libraries to create infrastructure of equivalent quality. To be fair to Amazon and other cloud providers, what I am asking for can be considered their competitive advantage, and on top of that would be incredibly difficult to open-source. In effect, it is not enough to have accessible APIs. A new cloud would need to be **architected** from the ground up to expose the primitives needed to create first class abstractions.
```

Comment:
This is a very good point to make but I would clarify on this a bit. What exactly are these results caused by? What can the AWS engineer do architecture/design/practices wise that I can't as an external engineer. If it's the work spent on libraries and experience with the low level infrastructure then state so. I'm just not so clear on it

>  It comes down to the architecture, the design and practices that AWS built over time to deliver multi-tenant infrastructure at scale. We can see the results in techniques such as shuffle sharding and static availability as well as designing services to be managed through control planes.

Reply:
The rest of the paragraph reads much clearer and is very good